### DISCUSSION TITLE
Collision behavior different for Custom Geometry

### DISCUSSION THREAD

**Discussion by yananZZZZ** [NVIDIA-Omniverse/PhysX#250] [Q&A] [answered]
Hi NVIDIA Devs,

As I try to use "PxCustomGeometryExt::CylinderCallbacks" to simulate cylinder collision with convex shapes. I notice that the behavior is rather different from convex VS convex. It looks like the the collision among convex is rather stable and accurate, which cause unexpected behavior for collision  in below use case. Can see from the second video, the custom cylinder seems pumping between the tunnel and get squeeze out eventually. however, when I change the cylinder into convex, it become much reasonable like the first one.
Is there any parameters or tolerance for custom geometry to tune this behavior? or anything related for me to try out?
![ConvexCollision](https://github.com/NVIDIA-Omniverse/PhysX/assets/122350362/de6c4932-d3d9-4dd7-bf5a-502a92df0edc)
 

![CylinderCollision1](https://github.com/NVIDIA-Omniverse/PhysX/assets/122350362/051e7a61-96d6-4323-a866-0e9d33552227)

**Comment by vreutskyy** [comment]
Hello yananZZZZ
Yes, sure, custom cylinder implementation isn't ideal, and its main disadvantage would probably be its performance. After all the original purpose the custom cylinder was made was to ensure it rolls smoothly atop a flat surface, which cannot be achieved with a convex mesh. For all other usages a built-in geometry like a primitive or a convex mesh would be preferable.
If you have a sample with the problematic setup we could try to tune it to make it more stable.
Thanks

**Reply by yananZZZZ** [reply]
Hi,
sorry for the late update, I managed to export an ovd file which can demo the custom cylinder unstable behavior with convex in Omniverse. Hope it can give an example how this cylinder been used in our application and maybe some setup advice from the expert.
[CustomCylinder.zip](https://github.com/NVIDIA-Omniverse/PhysX/files/15292235/CustomCylinder.zip)

thank you.

**Reply by vreutskyy** [reply]
Hello @yananZZZZ 
Thanks for the recording, we'll try to dig what's happening there.
I also wanted to ask if you have considered using a convex mesh instead of a cylinder?

Thanks

**Reply by yananZZZZ** [reply]
Hi Victor,

Thank you for the reply and indeed we use convex shape to simulate cylinder at first.  what's we did is using few facets to represent the cylindrical face and create convex shape however, it always not smooths enough as it should be.  Some cases this approximation even leads to wrong result. For instance, when we try to use raycast and get the reflect angle against cylinder. The angle varies in a big range even when the ray direction just changes a bit. 
We cannot figure out better solution for those kinds of issues. Hope to get some ideas from you.

Thank you

**Reply by vreutskyy** [reply]
May I ask what this is done for then? There can be different approaches to solve it, starting from the contact modification and up to making another custom geometry with the required behavior.

**Reply by yananZZZZ** [reply]
Hi Victor,

The use case of the cylinder is quite simple, just a dynamics actor with cylinder collision shape. But we want its scene query result such as "raycast" become more accurate. Using convex collision type we also encounter some explainable but unreasonable wrong result as i explain to you in the last reply. While, using custom geometry I cannot solve collision behavior for challenging shape. 

I believe improving convex shape scene query result is not a good way to go. But I'm not how to solve custom cylinder behavior issue using contact modification or other approach. as fact, I generally don't know why this behavior happens.

thank you.

**Reply by vreutskyy** [reply, marked as answer]
Can you add both shapes to the actor and set eSIMULATION_SHAPE for the convex mesh and eSCENE_QUERY_SHAPE for the cylinder?

**Reply by yananZZZZ** [reply]
Hi Victor,
You are right, this is our solution for now. It works for our application and the reason for this post is to let the NVIDIA experts explore if there is a improvement for the custom cylinder collision behavior because from our side there is no obvious explanation why this won't work compare to convex.

Thank you.

**Reply by vreutskyy** [reply]
Ah, I see. I just thought you didn't have a workaround for this issue. Otherwise, thanks for the report, we'll fix this in future versions.

Thanks
